

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">

  
  <title>Stability analysis of ICA components for transcriptomic data &mdash; stabilized-ica 1.0.1 documentation</title>
  
  <link rel="canonical" href="http://scikit-learn.org/stable/_examples/stability_study.html" />

  

  <link rel="stylesheet" href="../_static/css/vendor/bootstrap.min.css" type="text/css" />
  <link rel="stylesheet" href="../_static/gallery.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
<script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
<script src="../_static/jquery.js"></script> 
</head>
<body>
<nav id="navbar" class="sk-docs-navbar navbar navbar-expand-md navbar-light bg-light py-0">
  <div class="container-fluid sk-docs-container px-0">
      <a class="navbar-brand py-0" href="../index.html">
        <img
          class="sk-brand-img"
          src="../_static/stabilized-ica-logo_1.png"
          alt="logo"/>
      </a>
    <button
      id="sk-navbar-toggler"
      class="navbar-toggler"
      type="button"
      data-toggle="collapse"
      data-target="#navbarSupportedContent"
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
    >
      <span class="navbar-toggler-icon"></span>
    </button>

    <div class="sk-navbar-collapse collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav mr-auto">
        <li class="nav-item">
          <a class="sk-nav-link nav-link" href="../Install.html">Install</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link" href="../user_guide.html">User Guide</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link" href="../modules/classes.html">API</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link" href="../Examples.html">Examples</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link nav-more-item-mobile-items" href="../Release%20Notes.html">Release Notes</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link nav-more-item-mobile-items" href="../about.html">About us</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link nav-more-item-mobile-items" href="https://github.com/ncaptier/stabilized-ica">GitHub</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link nav-more-item-mobile-items" href="https://pypi.org/project/stabilized-ica/">PyPI</a>
        </li>
        <li class="nav-item dropdown nav-more-item-dropdown">
          <a class="sk-nav-link nav-link dropdown-toggle" href="#" id="navbarDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">More</a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdown">
              <a class="sk-nav-dropdown-item dropdown-item" href="../Release%20Notes.html">Release Notes</a>
              <a class="sk-nav-dropdown-item dropdown-item" href="../about.html">About us</a>
              <a class="sk-nav-dropdown-item dropdown-item" href="https://github.com/ncaptier/stabilized-ica">GitHub</a>
              <a class="sk-nav-dropdown-item dropdown-item" href="https://pypi.org/project/stabilized-ica/">PyPI</a>
          </div>
        </li>
      </ul>
      <div id="searchbox" role="search">
          <div class="searchformwrapper">
          <form class="search" action="../search.html" method="get">
            <input class="sk-search-text-input" type="text" name="q" aria-labelledby="searchlabel" />
            <input class="sk-search-text-btn" type="submit" value="Go" />
          </form>
          </div>
      </div>
    </div>
  </div>
</nav>
<div class="d-flex" id="sk-doc-wrapper">
    <input type="checkbox" name="sk-toggle-checkbox" id="sk-toggle-checkbox">
    <label id="sk-sidemenu-toggle" class="sk-btn-toggle-toc btn sk-btn-primary" for="sk-toggle-checkbox">Toggle Menu</label>
    <div id="sk-sidebar-wrapper" class="border-right">
      <div class="sk-sidebar-toc-wrapper">
        <div class="sk-sidebar-toc-logo">
          <a href="../index.html">
            <img
              class="sk-brand-img"
              src="../_static/stabilized-ica-logo_1.png"
              alt="logo"/>
          </a>
        </div>
        <div class="btn-group w-100 mb-2" role="group" aria-label="rellinks">
            <a href="#" role="button" class="btn sk-btn-rellink py-1 disabled"">Prev</a>
            <a href="#" role="button" class="btn sk-btn-rellink disabled py-1">Up</a>
            <a href="#" role="button" class="btn sk-btn-rellink py-1 disabled"">Next</a>
        </div>
        <div class="alert alert-danger p-1 mb-2" role="alert">
          <p class="text-center mb-0">
          <strong>stabilized-ica 1.0.1</strong><br/>
          <a href="../Release%20Notes.html">Release Notes</a>
          </p>
        </div>
        <div class="alert alert-warning p-1 mb-2" role="alert">
          <p class="text-center mb-0">
            Please <a class="font-weight-bold" href="../about.html#citing-stabilized-ica"><string>cite us</string></a> if you use the software.
          </p>
        </div>
            <div class="sk-sidebar-toc">
              <ul>
<li><a class="reference internal" href="#">Stability analysis of ICA components for transcriptomic data</a><ul>
<li><a class="reference internal" href="#0.-Load-data-sets">0. Load data sets</a><ul>
<li><a class="reference internal" href="#Grossman-data-set">Grossman data set</a></li>
<li><a class="reference internal" href="#Lim-NSCL-data-set">Lim NSCL data set</a></li>
</ul>
</li>
<li><a class="reference internal" href="#1.-Most-Stable-Transcriptome-Dimension-(MSTD)">1. Most Stable Transcriptome Dimension (MSTD)</a></li>
<li><a class="reference internal" href="#2.-Stabilized-ICA-decompostion-for-each-data-set">2. Stabilized ICA decompostion for each data set</a></li>
<li><a class="reference internal" href="#3.-Study-of-reproducibility---MNN-graph">3. Study of reproducibility - MNN graph</a></li>
<li><a class="reference internal" href="#4.-Cytoscape-visualization-of-the-stability-network">4. Cytoscape visualization of the stability network</a></li>
</ul>
</li>
</ul>

            </div>
      </div>
    </div>
    <div id="sk-page-content-wrapper">
      <div class="sk-page-content container-fluid body px-md-3" role="main">
        
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Stability-analysis-of-ICA-components-for-transcriptomic-data">
<h1>Stability analysis of ICA components for transcriptomic data<a class="headerlink" href="#Stability-analysis-of-ICA-components-for-transcriptomic-data" title="Permalink to this headline">¶</a></h1>
<p>Here we propose a short analysis of the stability and the reproductibility of ICA components extracted from several gene expression data sets. We are mainly interested in studying the behavior of transcriptomic data extracted from <a class="reference external" href="https://elifesciences.org/articles/23421">“Defining the Biological Basis of Radiomic Phenotypes in Lung Cancer” Grossman et al. 2017</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">autoreload</span> 2
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">time</span>
</pre></div>
</div>
</div>
<section id="0.-Load-data-sets">
<h2>0. Load data sets<a class="headerlink" href="#0.-Load-data-sets" title="Permalink to this headline">¶</a></h2>
<section id="Grossman-data-set">
<h3>Grossman data set<a class="headerlink" href="#Grossman-data-set" title="Permalink to this headline">¶</a></h3>
<p>These data sets were extracted from <a class="reference external" href="https://elifesciences.org/articles/23421">“Defining the Biological Basis of Radiomic Phenotypes in Lung Cancer” Grossman et al. 2017</a>.</p>
<p>df1 contains the expression of 21,766 unique genes for 269 patients with Non-small cell lung cancer (NSCLC) treated at the H. Lee Moffitt Cancer Center, Tampa, Florida, USA. df2 contains the expression of the same 21,766 unique genes for 89 patients with Non-small cell lung cancer (NSCLC)treated at MAASTRO clinical, Maastricht, NL. Gene expression values were measured on a custom Rosetta/Merck Affymetrix 2.0 microarray chipset and normalized with the robust multi-array average (RMA) algorithm.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">df1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s2">&quot;Data sets/data_USA.xlsx&quot;</span> <span class="p">,</span> <span class="n">sheet_name</span> <span class="o">=</span> <span class="s1">&#39;expression&#39;</span> <span class="p">,</span> <span class="n">index_col</span><span class="o">=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
<span class="n">df2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s2">&quot;Data sets/data_UE.xlsx&quot;</span> <span class="p">,</span> <span class="n">sheet_name</span> <span class="o">=</span> <span class="s1">&#39;expression&#39;</span> <span class="p">,</span> <span class="n">index_col</span><span class="o">=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>

<span class="c1">#With xlrd &gt;= 2.0.0 you will need to use a different engine since xrld removed support for anything</span>
<span class="c1">#other than .xls files (make sure openpyxl is installed). It will usually take a bit of time.</span>
<span class="c1">#df1 = pd.read_excel(&quot;Data sets/data_USA.xlsx&quot; , engine = &#39;openpyxl&#39; , sheet_name = &#39;expression&#39; , index_col= 0).transpose()</span>
<span class="c1">#df2 = pd.read_excel(&quot;Data sets/data_UE.xlsx&quot; ,  engine = &#39;openpyxl&#39; , sheet_name = &#39;expression&#39; , index_col= 0).transpose()</span>

<span class="n">dic</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df1</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">dic</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">col</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>

<span class="n">df1</span> <span class="o">=</span> <span class="n">df1</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">dic</span><span class="p">)</span>
<span class="n">df1</span><span class="p">[</span><span class="s1">&#39;Dataset&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Grossman USA&#39;</span><span class="p">]</span><span class="o">*</span><span class="n">df1</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">df1</span> <span class="o">=</span> <span class="n">df1</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;index&#39;</span> <span class="p">:</span> <span class="s1">&#39;Samples&#39;</span><span class="p">})</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s1">&#39;Dataset&#39;</span> <span class="p">,</span> <span class="s1">&#39;Samples&#39;</span><span class="p">])</span>

<span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">dic</span><span class="p">)</span>
<span class="n">df2</span><span class="p">[</span><span class="s1">&#39;Dataset&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Grossman UE&#39;</span><span class="p">]</span><span class="o">*</span><span class="n">df2</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;index&#39;</span> <span class="p">:</span> <span class="s1">&#39;Samples&#39;</span><span class="p">})</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s1">&#39;Dataset&#39;</span> <span class="p">,</span> <span class="s1">&#39;Samples&#39;</span><span class="p">])</span>

<span class="n">df1</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>3643</th>
      <th>84263</th>
      <th>7171</th>
      <th>2934</th>
      <th>11052</th>
      <th>1241</th>
      <th>6453</th>
      <th>57541</th>
      <th>9349</th>
      <th>11165</th>
      <th>...</th>
      <th>643669</th>
      <th>1572</th>
      <th>8551</th>
      <th>26784</th>
      <th>26783</th>
      <th>26782</th>
      <th>26779</th>
      <th>26778</th>
      <th>26777</th>
      <th>100132941</th>
    </tr>
    <tr>
      <th>Dataset</th>
      <th>Samples</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="5" valign="top">Grossman USA</th>
      <th>RadioGenomic-017</th>
      <td>5.205151</td>
      <td>7.097989</td>
      <td>9.559617</td>
      <td>8.396808</td>
      <td>7.603719</td>
      <td>7.990605</td>
      <td>10.044401</td>
      <td>9.054930</td>
      <td>7.383169</td>
      <td>8.177010</td>
      <td>...</td>
      <td>6.419273</td>
      <td>3.809826</td>
      <td>6.507880</td>
      <td>6.572121</td>
      <td>5.400848</td>
      <td>5.951391</td>
      <td>3.381860</td>
      <td>9.825584</td>
      <td>2.905091</td>
      <td>5.622438</td>
    </tr>
    <tr>
      <th>RadioGenomic-055</th>
      <td>5.615738</td>
      <td>6.585052</td>
      <td>9.777869</td>
      <td>9.082415</td>
      <td>8.639498</td>
      <td>6.781274</td>
      <td>9.541826</td>
      <td>8.866110</td>
      <td>6.422702</td>
      <td>7.196294</td>
      <td>...</td>
      <td>5.753828</td>
      <td>4.186127</td>
      <td>6.821582</td>
      <td>7.031406</td>
      <td>4.852417</td>
      <td>6.140850</td>
      <td>2.629760</td>
      <td>9.005145</td>
      <td>3.366466</td>
      <td>5.495330</td>
    </tr>
    <tr>
      <th>RadioGenomic-227</th>
      <td>5.679276</td>
      <td>7.747854</td>
      <td>10.648704</td>
      <td>9.127985</td>
      <td>7.369421</td>
      <td>7.203773</td>
      <td>8.972255</td>
      <td>8.328371</td>
      <td>7.269232</td>
      <td>7.449183</td>
      <td>...</td>
      <td>5.666999</td>
      <td>4.316130</td>
      <td>6.637855</td>
      <td>6.248824</td>
      <td>4.664228</td>
      <td>5.767970</td>
      <td>2.911470</td>
      <td>8.674466</td>
      <td>3.337194</td>
      <td>6.308605</td>
    </tr>
    <tr>
      <th>RadioGenomic-222</th>
      <td>5.317341</td>
      <td>7.196276</td>
      <td>10.949771</td>
      <td>8.098896</td>
      <td>7.639882</td>
      <td>7.971876</td>
      <td>10.159637</td>
      <td>8.667702</td>
      <td>8.474250</td>
      <td>7.271477</td>
      <td>...</td>
      <td>5.531060</td>
      <td>3.403776</td>
      <td>7.059419</td>
      <td>6.201873</td>
      <td>4.690005</td>
      <td>6.256286</td>
      <td>4.119688</td>
      <td>9.099659</td>
      <td>3.181781</td>
      <td>5.740033</td>
    </tr>
    <tr>
      <th>RadioGenomic-212</th>
      <td>7.196904</td>
      <td>9.346492</td>
      <td>9.673778</td>
      <td>9.358636</td>
      <td>8.741693</td>
      <td>7.616498</td>
      <td>10.376653</td>
      <td>8.701461</td>
      <td>6.601991</td>
      <td>7.344651</td>
      <td>...</td>
      <td>5.519642</td>
      <td>3.796049</td>
      <td>7.332635</td>
      <td>6.050121</td>
      <td>4.898523</td>
      <td>6.537895</td>
      <td>3.600895</td>
      <td>8.792510</td>
      <td>2.945391</td>
      <td>5.835411</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 21766 columns</p>
</div></div>
</div>
</section>
<section id="Lim-NSCL-data-set">
<h3>Lim NSCL data set<a class="headerlink" href="#Lim-NSCL-data-set" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">This data set was extracted from <a class="reference external" href="https://www.nature.com/articles/sdata2018136">“Data Descriptor: A merged lung cancer transcriptome dataset for clinical predictive modeling” Bin Lim et al. 2018</a>.</div>
<div class="line">It gathers 10 independent GEO cohorts of patients with Non-small cell lung cancer (NSCLC) or normal lung tissues. In total, it contains the expression of 10077 unique genes for 1118 patients with Non-small cell lung cancer (NSCLC) or normal lung tissues. The 10 GEO data sets use the same chip platform (Affymetrix Human Genome U133 Plus 2.0 Array). The expression values were normalized with the frozen Robust Multiarray Analysis algorithm (fRMA).</div>
</div>
<p><strong>Note :</strong> In order to be able to join this data set with the Grossman data set, we converted the official gene symbols to entrez ids. We used <a class="reference external" href="https://david.ncifcrf.gov/">the Database for Annotation, Visualization and Integrated Discovery(DAVID)</a></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">new_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s2">&quot;Data sets/new_data.xlsx&quot;</span> <span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;expression&#39;</span> <span class="p">,</span> <span class="s1">&#39;clinical&#39;</span><span class="p">]</span> <span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1">#With xlrd &gt;= 2.0.0 you will need to use a different engine since xrld removed support for anything</span>
<span class="c1">#other than .xls files (make sure openpyxl is installed). It will usually take a bit of time.</span>
<span class="c1">#new_data = pd.read_excel(&quot;Data sets/new_data.xlsx&quot; , engine = &#39;openpyxl&#39; , sheet_name=[&#39;expression&#39; , &#39;clinical&#39;] , index_col=0)</span>

<span class="n">new_data</span><span class="p">[</span><span class="s1">&#39;expression&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_data</span><span class="p">[</span><span class="s1">&#39;expression&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>

<span class="n">temp</span> <span class="o">=</span> <span class="n">new_data</span><span class="p">[</span><span class="s1">&#39;expression&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">new_data</span><span class="p">[</span><span class="s1">&#39;clinical&#39;</span><span class="p">][[</span><span class="s1">&#39;Dataset&#39;</span> <span class="p">,</span> <span class="s1">&#39;Histology (ADC: adenocarcinoma; LCC: large cell carcinoma; SCC: squamous cell carcinoma)&#39;</span><span class="p">]])</span>
<span class="n">df3</span> <span class="o">=</span> <span class="n">temp</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;index&#39;</span> <span class="p">:</span> <span class="s1">&#39;Samples&#39;</span><span class="p">})</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s1">&#39;Dataset&#39;</span> <span class="p">,</span> <span class="s1">&#39;Samples&#39;</span><span class="p">])</span>

<span class="k">del</span> <span class="n">temp</span> <span class="p">,</span> <span class="n">new_data</span>

<span class="n">df3</span> <span class="o">=</span> <span class="n">df3</span><span class="p">[</span><span class="n">df3</span><span class="p">[</span><span class="s1">&#39;Histology (ADC: adenocarcinoma; LCC: large cell carcinoma; SCC: squamous cell carcinoma)&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;Healthy&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="s1">&#39;Histology (ADC: adenocarcinoma; LCC: large cell carcinoma; SCC: squamous cell carcinoma)&#39;</span><span class="p">)</span>
<span class="n">df3</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>1</th>
      <th>2</th>
      <th>144571</th>
      <th>65985</th>
      <th>13</th>
      <th>201651</th>
      <th>51166</th>
      <th>195827</th>
      <th>79719</th>
      <th>22848</th>
      <th>...</th>
      <th>9183</th>
      <th>55055</th>
      <th>11130</th>
      <th>7789</th>
      <th>158586</th>
      <th>440590</th>
      <th>79699</th>
      <th>7791</th>
      <th>23140</th>
      <th>26009</th>
    </tr>
    <tr>
      <th>Dataset</th>
      <th>Samples</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="5" valign="top">GSE33356</th>
      <th>GSM494556</th>
      <td>5.234802</td>
      <td>11.922797</td>
      <td>5.031032</td>
      <td>9.141089</td>
      <td>9.934807</td>
      <td>6.868094</td>
      <td>5.038721</td>
      <td>6.845863</td>
      <td>7.877033</td>
      <td>9.201101</td>
      <td>...</td>
      <td>7.149422</td>
      <td>7.737312</td>
      <td>8.386004</td>
      <td>5.994052</td>
      <td>7.685643</td>
      <td>4.546110</td>
      <td>8.870484</td>
      <td>7.938569</td>
      <td>7.596789</td>
      <td>8.049103</td>
    </tr>
    <tr>
      <th>GSM494557</th>
      <td>5.069035</td>
      <td>11.146006</td>
      <td>4.737330</td>
      <td>8.926558</td>
      <td>4.419542</td>
      <td>3.276470</td>
      <td>7.517926</td>
      <td>6.945466</td>
      <td>7.387878</td>
      <td>9.279903</td>
      <td>...</td>
      <td>7.307666</td>
      <td>7.693829</td>
      <td>9.986055</td>
      <td>5.808656</td>
      <td>5.968874</td>
      <td>4.979624</td>
      <td>9.177783</td>
      <td>7.976926</td>
      <td>7.044507</td>
      <td>8.192161</td>
    </tr>
    <tr>
      <th>GSM494558</th>
      <td>5.514972</td>
      <td>12.378934</td>
      <td>6.990395</td>
      <td>7.268946</td>
      <td>4.253987</td>
      <td>2.978153</td>
      <td>6.286073</td>
      <td>7.201963</td>
      <td>7.125318</td>
      <td>10.170862</td>
      <td>...</td>
      <td>6.886556</td>
      <td>7.927103</td>
      <td>8.124558</td>
      <td>5.526909</td>
      <td>6.000892</td>
      <td>4.951939</td>
      <td>8.613027</td>
      <td>8.642108</td>
      <td>7.747791</td>
      <td>7.947040</td>
    </tr>
    <tr>
      <th>GSM494559</th>
      <td>6.871695</td>
      <td>11.853862</td>
      <td>4.589314</td>
      <td>8.874169</td>
      <td>5.911853</td>
      <td>4.412516</td>
      <td>6.304975</td>
      <td>7.731857</td>
      <td>7.300380</td>
      <td>8.649045</td>
      <td>...</td>
      <td>7.939107</td>
      <td>8.589020</td>
      <td>9.673297</td>
      <td>6.190925</td>
      <td>6.736635</td>
      <td>4.765782</td>
      <td>9.643763</td>
      <td>7.295604</td>
      <td>7.839339</td>
      <td>8.377581</td>
    </tr>
    <tr>
      <th>GSM494560</th>
      <td>6.761781</td>
      <td>11.200515</td>
      <td>5.001184</td>
      <td>9.027746</td>
      <td>5.356248</td>
      <td>3.515208</td>
      <td>5.545497</td>
      <td>7.552539</td>
      <td>7.309306</td>
      <td>9.229319</td>
      <td>...</td>
      <td>7.833734</td>
      <td>8.689572</td>
      <td>10.304345</td>
      <td>6.090843</td>
      <td>6.703918</td>
      <td>5.821131</td>
      <td>9.187349</td>
      <td>7.557380</td>
      <td>7.320086</td>
      <td>8.827688</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 10077 columns</p>
</div></div>
</div>
<p>We only keep data sets with more than 89 samples to avoid ending up with very noisy ICA components which will compromise the stability analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">dataframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span> <span class="p">,</span> <span class="n">df3</span><span class="p">],</span> <span class="n">join</span> <span class="o">=</span> <span class="s1">&#39;inner&#39;</span><span class="p">)</span>
<span class="n">names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">dataframe</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="mi">0</span><span class="p">))</span>

<span class="n">exclude</span> <span class="o">=</span> <span class="p">[]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Data sets in the study : &quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
    <span class="n">n_samples</span> <span class="o">=</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">n_samples</span> <span class="o">&lt;</span> <span class="mi">89</span><span class="p">:</span>
        <span class="n">exclude</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
       <span class="nb">print</span><span class="p">(</span><span class="n">name</span> <span class="p">,</span> <span class="n">n_samples</span><span class="p">)</span>

<span class="k">del</span> <span class="n">df1</span> <span class="p">,</span> <span class="n">df2</span> <span class="p">,</span> <span class="n">df3</span>

<span class="n">names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">names</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">exclude</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data sets in the study :
Grossman USA 262
Grossman UE 89
GSE19188 91
GSE28571 100
GSE50081 181
GSE31210 226
GSE18842 91
</pre></div></div>
</div>
</section>
</section>
<section id="1.-Most-Stable-Transcriptome-Dimension-(MSTD)">
<h2>1. Most Stable Transcriptome Dimension (MSTD)<a class="headerlink" href="#1.-Most-Stable-Transcriptome-Dimension-(MSTD)" title="Permalink to this headline">¶</a></h2>
<p>In order to select the number of ICA components we refer to <a class="reference external" href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-017-4112-9">“Determining the optimal number of independent components for reproducible transcriptomic data analysis” Kairov et al. 2017</a>. The idea is to find a trade-off between a sufficiently large number of components to capture all essential information and a restricted dimension to avoid ending up with a lot of very unstable components which only capture noise.</p>
<p>To do so we plot the stability distribution for a wide range of different number of components M = 5, … , 100 and we observe which choice of M most satisfies the trade-off the most.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#MSTD plot for the data set GSE31210.</span>
<span class="kn">from</span> <span class="nn">sica.base</span> <span class="kn">import</span> <span class="n">MSTD</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;GSE31210&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">func</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span> <span class="p">:</span> <span class="n">x</span> <span class="o">-</span> <span class="n">x</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
<span class="n">MSTD</span><span class="p">(</span><span class="n">X</span> <span class="p">,</span> <span class="mi">5</span> <span class="p">,</span> <span class="mi">100</span> <span class="p">,</span> <span class="mi">2</span> <span class="p">,</span> <span class="mi">20</span> <span class="p">,</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">3000</span><span class="p">)</span>
<span class="c1">#MSTD(X.values , 5 , 100 , 2 , 20 , max_iter = 3000)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "16f3a44a6ff947428ccc3a72381aabd6", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_examples_stability_study_9_2.png" src="../_images/_examples_stability_study_9_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">MSTD</span> <span class="o">=</span> <span class="mi">45</span>
</pre></div>
</div>
</div>
</section>
<section id="2.-Stabilized-ICA-decompostion-for-each-data-set">
<h2>2. Stabilized ICA decompostion for each data set<a class="headerlink" href="#2.-Stabilized-ICA-decompostion-for-each-data-set" title="Permalink to this headline">¶</a></h2>
<p>Our stabilized ICA algorithm is a copy of the ICASSO algorithm (implemented in MATLAB) (<a class="reference external" href="https://www.cs.helsinki.fi/u/ahyvarin/papers/Himberg03.pd">“ICASSO: software for investigating the reliability of ICA estimates by clustering and visualization” Himber et al.</a> The main idea is to iterate several times the FastICA algorithm (ex: from sklearn), cluster the results (agglomerative hierarchical clustering) and define the final ICA components as the centrotype of each cluster.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">sica.base</span> <span class="kn">import</span> <span class="n">StabilizedICA</span> <span class="k">as</span> <span class="n">sICA</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">Sources</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">decomp</span> <span class="o">=</span> <span class="n">sICA</span><span class="p">(</span><span class="n">n_components</span> <span class="o">=</span> <span class="n">MSTD</span> <span class="p">,</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">2000</span> <span class="p">,</span> <span class="n">n_jobs</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>

<span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">dataframe</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">func</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span> <span class="p">:</span> <span class="n">x</span> <span class="o">-</span> <span class="n">x</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="c1">#decomp.fit(X.values , n_runs = 30)</span>
    <span class="n">decomp</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span> <span class="p">,</span> <span class="n">n_runs</span> <span class="o">=</span> <span class="mi">30</span><span class="p">)</span>
    <span class="n">Sources</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">decomp</span><span class="o">.</span><span class="n">S_</span> <span class="p">)</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">minutes</span><span class="p">,</span> <span class="n">seconds</span> <span class="o">=</span> <span class="nb">divmod</span><span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">,</span> <span class="mi">60</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot; done !&quot;</span> <span class="p">,</span> <span class="s2">&quot;running time (min): &quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="si">{:0&gt;2}</span><span class="s2">:</span><span class="si">{:05.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">minutes</span><span class="p">),</span><span class="n">seconds</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Grossman USA done ! running time (min): 00:25.47
GSE50081 done ! running time (min): 00:14.74
GSE28571 done ! running time (min): 00:17.53
GSE19188 done ! running time (min): 00:16.74
GSE31210 done ! running time (min): 00:14.51
GSE18842 done ! running time (min): 00:16.21
Grossman UE done ! running time (min): 00:15.11
</pre></div></div>
</div>
</section>
<section id="3.-Study-of-reproducibility---MNN-graph">
<h2>3. Study of reproducibility - MNN graph<a class="headerlink" href="#3.-Study-of-reproducibility---MNN-graph" title="Permalink to this headline">¶</a></h2>
<p>In order to study the reproducibility of ICA components between the four data sets we use the notion of Reciprocal Best Hit (RBH). We consider that a component <span class="math notranslate nohighlight">\(i\)</span> extracted from a data set D1 and another component <span class="math notranslate nohighlight">\(j\)</span> extracted from a data set D2 are linked if:</p>
<p><span class="math">\begin{equation}
|\rho_{ij}| = \max \{ |\rho_{ik}| , \forall k \, \in D2 \} = \max \{ |\rho_{lj}| , \forall l \, \in D1 \}
\end{equation}</span></p>
<p>Here <span class="math notranslate nohighlight">\(\rho_{ij}\)</span> represents the Pearson’s correlation coefficient between the components <span class="math notranslate nohighlight">\(i\)</span> and <span class="math notranslate nohighlight">\(j\)</span>. We chose to use Pearson’s coefficient over non-parametric coefficients such as Kendall’s or Spearman’s coefficients since it is strongly influenced by extreme values. Yet in our case these extreme values characterize our ICA components .</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">sica.mutualknn</span> <span class="kn">import</span> <span class="n">MNNgraph</span>
</pre></div>
</div>
</div>
<p><strong>Note :</strong> The layout of the following RBH network is strongly influenced by a random initialisation (Fruchterman-Reingold force-directed algorithm). One will have to run the following result several times until a satisfying result is obtained.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fig</span> <span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">10</span> <span class="p">,</span> <span class="mi">7</span><span class="p">))</span>

<span class="n">cg</span> <span class="o">=</span> <span class="n">MNNgraph</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">Sources</span> <span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="n">names</span> <span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span> <span class="p">,</span> <span class="n">weighted</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">cg</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;r&#39;</span> <span class="p">,</span> <span class="s1">&#39;k&#39;</span> <span class="p">,</span> <span class="s1">&#39;b&#39;</span> <span class="p">,</span> <span class="s1">&#39;g&#39;</span> <span class="p">,</span> <span class="s1">&#39;yellow&#39;</span> <span class="p">,</span> <span class="s1">&#39;purple&#39;</span> <span class="p">,</span> <span class="s1">&#39;brown&#39;</span><span class="p">]</span> <span class="p">,</span> <span class="n">spacing</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.25</span> <span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x1ae423ad130&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_examples_stability_study_17_1.png" src="../_images/_examples_stability_study_17_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cg</span><span class="o">.</span><span class="n">export_json</span><span class="p">(</span><span class="s2">&quot;stability.json&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We can also draw the bipartite RBH graph for the ICA components of two data sets. We display the Pearson correlation coefficient for each link. In the following we display the bipartite graphs for the couples of sets (‘Grossman UE’ , ‘GSE19188’) and (‘Grossman UE’ , ‘GSE50081’).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fig</span> <span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span> <span class="p">,</span> <span class="mi">2</span> <span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span> <span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
    <span class="n">cg</span> <span class="o">=</span> <span class="n">MNNgraph</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">Sources</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="p">,</span> <span class="n">Sources</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]]</span> <span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="p">[</span><span class="n">names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="p">,</span> <span class="n">names</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]]</span> <span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">1</span> <span class="p">,</span> <span class="n">weighted</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">cg</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="p">,</span> <span class="n">bipartite_graph</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_examples_stability_study_20_0.png" src="../_images/_examples_stability_study_20_0.png" />
</div>
</div>
</section>
<section id="4.-Cytoscape-visualization-of-the-stability-network">
<h2>4. Cytoscape visualization of the stability network<a class="headerlink" href="#4.-Cytoscape-visualization-of-the-stability-network" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">In order to visualize the stability network with cytoscape directly on the jupyter notebook we use the <a class="reference external" href="https://github.com/QuantStack/ipycytoscape">ipycytoscape package</a>. Please install it with:</div>
<div class="line">* conda install -c conda-forge ipycytoscape * pip install ipycytoscape</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">ipycytoscape</span>
<span class="kn">import</span> <span class="nn">json</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;stability.json&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fi</span><span class="p">:</span>
    <span class="n">json_file</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">fi</span><span class="p">)</span>

<span class="n">cytoscapeobj</span> <span class="o">=</span> <span class="n">ipycytoscape</span><span class="o">.</span><span class="n">CytoscapeWidget</span><span class="p">()</span>
<span class="n">cytoscapeobj</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">add_graph_from_json</span><span class="p">(</span><span class="n">json_file</span><span class="p">[</span><span class="s1">&#39;elements&#39;</span><span class="p">])</span>

<span class="n">cytoscapeobj</span><span class="o">.</span><span class="n">set_layout</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;cose&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cytoscapeobj</span><span class="o">.</span><span class="n">set_style</span><span class="p">([{</span><span class="s1">&#39;selector&#39;</span><span class="p">:</span> <span class="s1">&#39;node&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;css&#39;</span><span class="p">:</span> <span class="p">{</span>
                                  <span class="s1">&#39;content&#39;</span><span class="p">:</span> <span class="s1">&#39;data(name)&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;font-size&#39;</span><span class="p">:</span> <span class="s1">&#39;12px&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;text-valign&#39;</span><span class="p">:</span> <span class="s1">&#39;center&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;text-halign&#39;</span><span class="p">:</span> <span class="s1">&#39;center&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;background-color&#39;</span><span class="p">:</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;b&#39;</span><span class="p">,</span>
                                <span class="p">}</span>
                        <span class="p">},</span>
                        <span class="p">{</span><span class="s1">&#39;selector&#39;</span><span class="p">:</span> <span class="s1">&#39;edge&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;css&#39;</span><span class="p">:</span> <span class="p">{</span>
                                  <span class="s1">&#39;curve-style&#39;</span><span class="p">:</span> <span class="s1">&#39;haystack&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;haystack-radius&#39;</span><span class="p">:</span> <span class="s1">&#39;0.5&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;opacity&#39;</span><span class="p">:</span> <span class="s1">&#39;0.7&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;line-color&#39;</span><span class="p">:</span> <span class="s1">&#39;#bbb&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;width&#39;</span><span class="p">:</span> <span class="s1">&#39;mapData(weight , 0 , 1 , 1 , 15)&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;overlay-padding&#39;</span><span class="p">:</span> <span class="s1">&#39;3px&#39;</span>
                                <span class="p">}</span>
                        <span class="p">}])</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cytoscapeobj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "2284a0e4b69347fcb9ed2bab8b2b0aad", "version_major": 2, "version_minor": 0}</script></div>
</div>
</section>
</section>


      </div>
    <div class="container">
      <footer class="sk-content-footer">
            &copy; 2007-2021 The scikit-learn developers (BSD License).
          <a href="../_sources/_examples/stability_study.ipynb.txt" rel="nofollow">Show this page source</a>
      </footer>
    </div>
  </div>
</div>
<script src="../_static/js/vendor/bootstrap.min.js"></script>

<script>
    window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
    ga('create', 'UA-22606712-2', 'auto');
    ga('set', 'anonymizeIp', true);
    ga('send', 'pageview');
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>


<script>
$(document).ready(function() {
    /* Add a [>>>] button on the top-right corner of code samples to hide
     * the >>> and ... prompts and the output and thus make the code
     * copyable. */
    var div = $('.highlight-python .highlight,' +
                '.highlight-python3 .highlight,' +
                '.highlight-pycon .highlight,' +
		'.highlight-default .highlight')
    var pre = div.find('pre');

    // get the styles from the current theme
    pre.parent().parent().css('position', 'relative');
    var hide_text = 'Hide prompts and outputs';
    var show_text = 'Show prompts and outputs';

    // create and add the button to all the code blocks that contain >>>
    div.each(function(index) {
        var jthis = $(this);
        if (jthis.find('.gp').length > 0) {
            var button = $('<span class="copybutton">&gt;&gt;&gt;</span>');
            button.attr('title', hide_text);
            button.data('hidden', 'false');
            jthis.prepend(button);
        }
        // tracebacks (.gt) contain bare text elements that need to be
        // wrapped in a span to work with .nextUntil() (see later)
        jthis.find('pre:has(.gt)').contents().filter(function() {
            return ((this.nodeType == 3) && (this.data.trim().length > 0));
        }).wrap('<span>');
    });

    // define the behavior of the button when it's clicked
    $('.copybutton').click(function(e){
        e.preventDefault();
        var button = $(this);
        if (button.data('hidden') === 'false') {
            // hide the code output
            button.parent().find('.go, .gp, .gt').hide();
            button.next('pre').find('.gt').nextUntil('.gp, .go').css('visibility', 'hidden');
            button.css('text-decoration', 'line-through');
            button.attr('title', show_text);
            button.data('hidden', 'true');
        } else {
            // show the code output
            button.parent().find('.go, .gp, .gt').show();
            button.next('pre').find('.gt').nextUntil('.gp, .go').css('visibility', 'visible');
            button.css('text-decoration', 'none');
            button.attr('title', hide_text);
            button.data('hidden', 'false');
        }
    });

	/*** Add permalink buttons next to glossary terms ***/
	$('dl.glossary > dt[id]').append(function() {
		return ('<a class="headerlink" href="#' +
			    this.getAttribute('id') +
			    '" title="Permalink to this term">¶</a>');
	});
  /*** Hide navbar when scrolling down ***/
  // Returns true when headerlink target matches hash in url
  (function() {
    hashTargetOnTop = function() {
        var hash = window.location.hash;
        if ( hash.length < 2 ) { return false; }

        var target = document.getElementById( hash.slice(1) );
        if ( target === null ) { return false; }

        var top = target.getBoundingClientRect().top;
        return (top < 2) && (top > -2);
    };

    // Hide navbar on load if hash target is on top
    var navBar = document.getElementById("navbar");
    var navBarToggler = document.getElementById("sk-navbar-toggler");
    var navBarHeightHidden = "-" + navBar.getBoundingClientRect().height + "px";
    var $window = $(window);

    hideNavBar = function() {
        navBar.style.top = navBarHeightHidden;
    };

    showNavBar = function() {
        navBar.style.top = "0";
    }

    if (hashTargetOnTop()) {
        hideNavBar()
    }

    var prevScrollpos = window.pageYOffset;
    hideOnScroll = function(lastScrollTop) {
        if (($window.width() < 768) && (navBarToggler.getAttribute("aria-expanded") === 'true')) {
            return;
        }
        if (lastScrollTop > 2 && (prevScrollpos <= lastScrollTop) || hashTargetOnTop()){
            hideNavBar()
        } else {
            showNavBar()
        }
        prevScrollpos = lastScrollTop;
    };

    /*** high performance scroll event listener***/
    var raf = window.requestAnimationFrame ||
        window.webkitRequestAnimationFrame ||
        window.mozRequestAnimationFrame ||
        window.msRequestAnimationFrame ||
        window.oRequestAnimationFrame;
    var lastScrollTop = $window.scrollTop();

    if (raf) {
        loop();
    }

    function loop() {
        var scrollTop = $window.scrollTop();
        if (lastScrollTop === scrollTop) {
            raf(loop);
            return;
        } else {
            lastScrollTop = scrollTop;
            hideOnScroll(lastScrollTop);
            raf(loop);
        }
    }
  })();
});

</script>
    
</body>
</html>