

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">

  
  <title>Stabilized ICA for transcriptomic data &mdash; stabilized-ica 1.0.1 documentation</title>
  
  <link rel="canonical" href="http://scikit-learn.org/stable/_examples/transcriptomic_ICA.html" />

  

  <link rel="stylesheet" href="../_static/css/vendor/bootstrap.min.css" type="text/css" />
  <link rel="stylesheet" href="../_static/gallery.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
<script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
<script src="../_static/jquery.js"></script> 
</head>
<body>
<nav id="navbar" class="sk-docs-navbar navbar navbar-expand-md navbar-light bg-light py-0">
  <div class="container-fluid sk-docs-container px-0">
      <a class="navbar-brand py-0" href="../index.html">
        <img
          class="sk-brand-img"
          src="../_static/stabilized-ica-logo_1.png"
          alt="logo"/>
      </a>
    <button
      id="sk-navbar-toggler"
      class="navbar-toggler"
      type="button"
      data-toggle="collapse"
      data-target="#navbarSupportedContent"
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
    >
      <span class="navbar-toggler-icon"></span>
    </button>

    <div class="sk-navbar-collapse collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav mr-auto">
        <li class="nav-item">
          <a class="sk-nav-link nav-link" href="../Install.html">Install</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link" href="../user_guide.html">User Guide</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link" href="../modules/classes.html">API</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link" href="../Examples.html">Examples</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link nav-more-item-mobile-items" href="../Release%20Notes.html">Release Notes</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link nav-more-item-mobile-items" href="../about.html">About us</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link nav-more-item-mobile-items" href="https://github.com/ncaptier/stabilized-ica">GitHub</a>
        </li>
        <li class="nav-item">
          <a class="sk-nav-link nav-link nav-more-item-mobile-items" href="https://pypi.org/project/stabilized-ica/">PyPI</a>
        </li>
        <li class="nav-item dropdown nav-more-item-dropdown">
          <a class="sk-nav-link nav-link dropdown-toggle" href="#" id="navbarDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">More</a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdown">
              <a class="sk-nav-dropdown-item dropdown-item" href="../Release%20Notes.html">Release Notes</a>
              <a class="sk-nav-dropdown-item dropdown-item" href="../about.html">About us</a>
              <a class="sk-nav-dropdown-item dropdown-item" href="https://github.com/ncaptier/stabilized-ica">GitHub</a>
              <a class="sk-nav-dropdown-item dropdown-item" href="https://pypi.org/project/stabilized-ica/">PyPI</a>
          </div>
        </li>
      </ul>
      <div id="searchbox" role="search">
          <div class="searchformwrapper">
          <form class="search" action="../search.html" method="get">
            <input class="sk-search-text-input" type="text" name="q" aria-labelledby="searchlabel" />
            <input class="sk-search-text-btn" type="submit" value="Go" />
          </form>
          </div>
      </div>
    </div>
  </div>
</nav>
<div class="d-flex" id="sk-doc-wrapper">
    <input type="checkbox" name="sk-toggle-checkbox" id="sk-toggle-checkbox">
    <label id="sk-sidemenu-toggle" class="sk-btn-toggle-toc btn sk-btn-primary" for="sk-toggle-checkbox">Toggle Menu</label>
    <div id="sk-sidebar-wrapper" class="border-right">
      <div class="sk-sidebar-toc-wrapper">
        <div class="sk-sidebar-toc-logo">
          <a href="../index.html">
            <img
              class="sk-brand-img"
              src="../_static/stabilized-ica-logo_1.png"
              alt="logo"/>
          </a>
        </div>
        <div class="btn-group w-100 mb-2" role="group" aria-label="rellinks">
            <a href="#" role="button" class="btn sk-btn-rellink py-1 disabled"">Prev</a>
            <a href="#" role="button" class="btn sk-btn-rellink disabled py-1">Up</a>
            <a href="#" role="button" class="btn sk-btn-rellink py-1 disabled"">Next</a>
        </div>
        <div class="alert alert-danger p-1 mb-2" role="alert">
          <p class="text-center mb-0">
          <strong>stabilized-ica 1.0.1</strong><br/>
          <a href="../Release%20Notes.html">Release Notes</a>
          </p>
        </div>
        <div class="alert alert-warning p-1 mb-2" role="alert">
          <p class="text-center mb-0">
            Please <a class="font-weight-bold" href="../about.html#citing-stabilized-ica"><string>cite us</string></a> if you use the software.
          </p>
        </div>
            <div class="sk-sidebar-toc">
              <ul>
<li><a class="reference internal" href="#">Stabilized ICA for transcriptomic data</a><ul>
<li><a class="reference internal" href="#0.-Load-the-data-set">0. Load the data set</a></li>
<li><a class="reference internal" href="#1.-Most-Stable-Transcriptome-Dimension-(MSTD)">1. Most Stable Transcriptome Dimension (MSTD)</a></li>
<li><a class="reference internal" href="#2.-Stabilized-ICA-and-Multidimensional-Scaling-(MDS)">2. Stabilized ICA and Multidimensional Scaling (MDS)</a></li>
<li><a class="reference internal" href="#3.-Metagene-annotation">3. Metagene annotation</a><ul>
<li><a class="reference internal" href="#3.1-Reactome-enrichment-analysis">3.1 Reactome enrichment analysis</a></li>
<li><a class="reference internal" href="#3.2-Toppfun-enrichment-analysis">3.2 Toppfun enrichment analysis</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            </div>
      </div>
    </div>
    <div id="sk-page-content-wrapper">
      <div class="sk-page-content container-fluid body px-md-3" role="main">
        
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Stabilized-ICA-for-transcriptomic-data">
<h1>Stabilized ICA for transcriptomic data<a class="headerlink" href="#Stabilized-ICA-for-transcriptomic-data" title="Permalink to this headline">¶</a></h1>
<p>In this jupyter notebook, we propose a short example for the use of our stabilized ICA algorithm. We study the extraction of ICA components for a transcriptomic data set gathering NSCLC patients.</p>
<p><strong>Note :</strong> To run this jupyter notebook, you first need to extract the data set “data.csv” from the zip file “data.zip” (same directory)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">autoreload</span> 2

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">time</span>
</pre></div>
</div>
</div>
<section id="0.-Load-the-data-set">
<h2>0. Load the data set<a class="headerlink" href="#0.-Load-the-data-set" title="Permalink to this headline">¶</a></h2>
<div class="line-block">
<div class="line">This data set was extraced from <a class="reference external" href="https://elifesciences.org/articles/23421">“Defining the Biological Basis of Radiomic Phenotypes in Lung Cancer” Grossman et al. 2017</a>.</div>
<div class="line">It contains the expression of 21,766 unique genes for 269 patients with Non-small cell lung cancer (NSCLC) treated at the H. Lee Moffitt Cancer Center, Tampa, Florida, USA. Gene expression values were measured on a custom Rosetta/Merck Affymetrix 2.0 microarray chipset and normalized with the robust multi-array average (RMA) algorithm.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;data.csv&quot;</span> <span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>RadioGenomic-017</th>
      <th>RadioGenomic-055</th>
      <th>RadioGenomic-227</th>
      <th>RadioGenomic-222</th>
      <th>RadioGenomic-212</th>
      <th>RadioGenomic-191</th>
      <th>RadioGenomic-259</th>
      <th>RadioGenomic-110</th>
      <th>RadioGenomic-003</th>
      <th>RadioGenomic-034</th>
      <th>...</th>
      <th>RadioGenomic-261</th>
      <th>RadioGenomic-204</th>
      <th>RadioGenomic-084</th>
      <th>RadioGenomic-232</th>
      <th>RadioGenomic-049</th>
      <th>RadioGenomic-186</th>
      <th>RadioGenomic-134</th>
      <th>RadioGenomic-239</th>
      <th>RadioGenomic-162</th>
      <th>RadioGenomic-179</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3643</th>
      <td>5.205151</td>
      <td>5.615738</td>
      <td>5.679276</td>
      <td>5.317341</td>
      <td>7.196904</td>
      <td>5.405730</td>
      <td>5.218404</td>
      <td>5.607431</td>
      <td>6.544434</td>
      <td>5.848995</td>
      <td>...</td>
      <td>5.115181</td>
      <td>5.732982</td>
      <td>5.713161</td>
      <td>5.922955</td>
      <td>5.796744</td>
      <td>5.620590</td>
      <td>5.435381</td>
      <td>7.151574</td>
      <td>6.706473</td>
      <td>6.536985</td>
    </tr>
    <tr>
      <th>84263</th>
      <td>7.097989</td>
      <td>6.585052</td>
      <td>7.747854</td>
      <td>7.196276</td>
      <td>9.346492</td>
      <td>7.662668</td>
      <td>7.438435</td>
      <td>8.463022</td>
      <td>7.952210</td>
      <td>7.936425</td>
      <td>...</td>
      <td>7.855756</td>
      <td>7.651911</td>
      <td>7.621733</td>
      <td>7.889039</td>
      <td>8.413899</td>
      <td>8.264302</td>
      <td>7.346372</td>
      <td>7.480408</td>
      <td>7.967057</td>
      <td>7.924989</td>
    </tr>
    <tr>
      <th>7171</th>
      <td>9.559617</td>
      <td>9.777869</td>
      <td>10.648704</td>
      <td>10.949771</td>
      <td>9.673778</td>
      <td>10.668188</td>
      <td>9.227177</td>
      <td>9.696762</td>
      <td>10.185486</td>
      <td>10.049003</td>
      <td>...</td>
      <td>9.742841</td>
      <td>10.102686</td>
      <td>9.533121</td>
      <td>11.825210</td>
      <td>10.096616</td>
      <td>10.297810</td>
      <td>10.160288</td>
      <td>10.145422</td>
      <td>8.893938</td>
      <td>11.080574</td>
    </tr>
    <tr>
      <th>2934</th>
      <td>8.396808</td>
      <td>9.082415</td>
      <td>9.127985</td>
      <td>8.098896</td>
      <td>9.358636</td>
      <td>8.421808</td>
      <td>8.643859</td>
      <td>8.445746</td>
      <td>9.233013</td>
      <td>9.350872</td>
      <td>...</td>
      <td>7.504968</td>
      <td>7.035265</td>
      <td>8.551597</td>
      <td>8.271226</td>
      <td>6.998375</td>
      <td>9.346116</td>
      <td>8.283634</td>
      <td>7.095790</td>
      <td>8.356098</td>
      <td>8.978290</td>
    </tr>
    <tr>
      <th>11052</th>
      <td>7.603719</td>
      <td>8.639498</td>
      <td>7.369421</td>
      <td>7.639882</td>
      <td>8.741693</td>
      <td>8.306786</td>
      <td>6.706823</td>
      <td>6.024337</td>
      <td>6.840010</td>
      <td>8.427551</td>
      <td>...</td>
      <td>6.497460</td>
      <td>8.156322</td>
      <td>8.075888</td>
      <td>7.541231</td>
      <td>8.135110</td>
      <td>7.826186</td>
      <td>7.820982</td>
      <td>6.427032</td>
      <td>8.622156</td>
      <td>7.127454</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 262 columns</p>
</div></div>
</div>
<p><strong>Note :</strong> in the following step, we center the expression of each gene (i.e the rows of the dataframe df). Indeed, we noticed that omitting this pre-processing step led to one ICA component entirely driven by the differences between the mean expression of each gene. This component is not biologically relevant and we avoid it thanks to this centering step.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">X</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">func</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span> <span class="p">:</span> <span class="n">x</span> <span class="o">-</span> <span class="n">x</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="1.-Most-Stable-Transcriptome-Dimension-(MSTD)">
<h2>1. Most Stable Transcriptome Dimension (MSTD)<a class="headerlink" href="#1.-Most-Stable-Transcriptome-Dimension-(MSTD)" title="Permalink to this headline">¶</a></h2>
<p>In order to select the number of ICA components we refer to <a class="reference external" href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-017-4112-9">“Determining the optimal number of independent components for reproducible transcriptomic data analysis” Kairov et al. 2017</a>. The idea is to find a trade-off between a sufficiently large number of components to capture all essential information and a restricted dimension to avoid ending up with a lot of very unstable components which only capture noise.</p>
<p>To do so we plot the stability distribution for a wide range of different number of components M = 5, … , 100 and we observe which choice of M most satisfies the trade-off.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">sica.base</span> <span class="kn">import</span> <span class="n">MSTD</span>
<span class="n">MSTD</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">values</span> <span class="p">,</span> <span class="n">m</span> <span class="o">=</span> <span class="mi">5</span> <span class="p">,</span> <span class="n">M</span> <span class="o">=</span> <span class="mi">100</span> <span class="p">,</span> <span class="n">step</span> <span class="o">=</span> <span class="mi">2</span> <span class="p">,</span> <span class="n">n_runs</span> <span class="o">=</span> <span class="mi">20</span> <span class="p">,</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "186f56b88f0f4c50bb1820dc2bf9feef", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_examples_transcriptomic_ICA_7_2.png" src="../_images/_examples_transcriptomic_ICA_7_2.png" />
</div>
</div>
<p>Please note that in the following we use a rough estimation of the optimal dimension (observed thanks to the graphs above). In fact more sophisticated techniques can be applied to estimate this dimension (ex: two-lines clustering) ; these can be found in the article cited above.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">dim_opt</span> <span class="o">=</span> <span class="mi">45</span>
</pre></div>
</div>
</div>
</section>
<section id="2.-Stabilized-ICA-and-Multidimensional-Scaling-(MDS)">
<h2>2. Stabilized ICA and Multidimensional Scaling (MDS)<a class="headerlink" href="#2.-Stabilized-ICA-and-Multidimensional-Scaling-(MDS)" title="Permalink to this headline">¶</a></h2>
<p>This stabilized ICA algorithm is a copy of the ICASSO algorithm (implemented in MATLAB) (<a class="reference external" href="https://www.cs.helsinki.fi/u/ahyvarin/papers/Himberg03.pd">“ICASSO: software for investigating the reliability of ICA estimates by clustering and visualization” Himber et al.</a> The main idea is to iterate several times the FastICA algorithm (ex: from sklearn), cluster the results (agglomerative hierarchical clustering) and define the final ICA components as the centrotype of each cluster.</p>
<p>Some explanation for Multidimensional Scaling can be found on the sklearn page : <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.manifold.MDS.html">https://scikit-learn.org/stable/modules/generated/sklearn.manifold.MDS.html</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">sica.base</span> <span class="kn">import</span> <span class="n">StabilizedICA</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

<span class="n">sICA</span> <span class="o">=</span> <span class="n">StabilizedICA</span><span class="p">(</span><span class="n">n_components</span> <span class="o">=</span> <span class="n">dim_opt</span> <span class="p">,</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">2000</span> <span class="p">,</span> <span class="n">n_jobs</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="n">sICA</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span> <span class="p">,</span> <span class="n">n_runs</span> <span class="o">=</span> <span class="mi">30</span> <span class="p">,</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span> <span class="p">,</span><span class="n">normalize</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

<span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

<span class="n">minutes</span><span class="p">,</span> <span class="n">seconds</span> <span class="o">=</span> <span class="nb">divmod</span><span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">,</span> <span class="mi">60</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;running time (min): &quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="si">{:0&gt;2}</span><span class="s2">:</span><span class="si">{:05.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">minutes</span><span class="p">),</span><span class="n">seconds</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
running time (min): 00:29.84
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_examples_transcriptomic_ICA_12_1.png" src="../_images/_examples_transcriptomic_ICA_12_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fig</span> <span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">10</span> <span class="p">,</span> <span class="mi">7</span><span class="p">))</span>

<span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

<span class="n">sICA</span><span class="o">.</span><span class="n">projection</span><span class="p">(</span><span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span> <span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;mds&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s2">&quot;Multidimensional scaling for ICA components&quot;</span><span class="p">)</span>

<span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

<span class="n">minutes</span><span class="p">,</span> <span class="n">seconds</span> <span class="o">=</span> <span class="nb">divmod</span><span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">,</span> <span class="mi">60</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;running time (min): &quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="si">{:0&gt;2}</span><span class="s2">:</span><span class="si">{:05.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">minutes</span><span class="p">),</span><span class="n">seconds</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
running time (min): 00:28.23
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_examples_transcriptomic_ICA_13_1.png" src="../_images/_examples_transcriptomic_ICA_13_1.png" />
</div>
</div>
<p><strong>Note :</strong> by default, each stabilized ICA component (i.e a row of the S_ matrix) is normalized to unit variance and re-oriented towards positive heaviest tail. These operations facilitate the interpretation of the components and allow us to compare several components from different studies.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">Metagenes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">sICA</span><span class="o">.</span><span class="n">S_</span> <span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;metagene &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">sICA</span><span class="o">.</span><span class="n">S_</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])])</span>
<span class="n">Metagenes</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>3643</th>
      <th>84263</th>
      <th>7171</th>
      <th>2934</th>
      <th>11052</th>
      <th>1241</th>
      <th>6453</th>
      <th>57541</th>
      <th>9349</th>
      <th>11165</th>
      <th>...</th>
      <th>643669</th>
      <th>1572</th>
      <th>8551</th>
      <th>26784</th>
      <th>26783</th>
      <th>26782</th>
      <th>26779</th>
      <th>26778</th>
      <th>26777</th>
      <th>100132941</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>metagene 0</th>
      <td>-0.579485</td>
      <td>0.071363</td>
      <td>0.616933</td>
      <td>-0.643553</td>
      <td>-0.068441</td>
      <td>-0.545932</td>
      <td>0.026524</td>
      <td>0.225038</td>
      <td>-0.411407</td>
      <td>0.128290</td>
      <td>...</td>
      <td>0.155766</td>
      <td>-0.373558</td>
      <td>-1.049004</td>
      <td>-0.100770</td>
      <td>0.288055</td>
      <td>0.115193</td>
      <td>-0.667923</td>
      <td>-0.191776</td>
      <td>-0.029620</td>
      <td>-0.289703</td>
    </tr>
    <tr>
      <th>metagene 1</th>
      <td>0.474153</td>
      <td>-0.121899</td>
      <td>-0.176296</td>
      <td>0.481017</td>
      <td>0.336107</td>
      <td>0.712320</td>
      <td>0.427755</td>
      <td>-0.426953</td>
      <td>0.695701</td>
      <td>0.019340</td>
      <td>...</td>
      <td>-0.290772</td>
      <td>3.189300</td>
      <td>-0.132017</td>
      <td>-0.176921</td>
      <td>-0.316637</td>
      <td>-0.149949</td>
      <td>-0.444736</td>
      <td>0.236175</td>
      <td>0.320990</td>
      <td>-0.018504</td>
    </tr>
    <tr>
      <th>metagene 2</th>
      <td>-0.732185</td>
      <td>-0.228899</td>
      <td>0.112325</td>
      <td>0.007528</td>
      <td>1.312248</td>
      <td>0.188654</td>
      <td>-0.245325</td>
      <td>-0.132094</td>
      <td>-2.309839</td>
      <td>0.882917</td>
      <td>...</td>
      <td>-0.630225</td>
      <td>-0.781150</td>
      <td>-0.453338</td>
      <td>-0.027245</td>
      <td>0.137956</td>
      <td>1.619909</td>
      <td>0.443488</td>
      <td>0.654490</td>
      <td>-0.039455</td>
      <td>-0.383547</td>
    </tr>
    <tr>
      <th>metagene 3</th>
      <td>-0.034504</td>
      <td>0.032134</td>
      <td>0.564410</td>
      <td>1.366289</td>
      <td>-0.290005</td>
      <td>2.460954</td>
      <td>3.451236</td>
      <td>0.313231</td>
      <td>-1.121460</td>
      <td>1.434259</td>
      <td>...</td>
      <td>0.620856</td>
      <td>0.362091</td>
      <td>0.578978</td>
      <td>-1.417301</td>
      <td>-1.061770</td>
      <td>0.479641</td>
      <td>0.677915</td>
      <td>-0.628650</td>
      <td>-0.362539</td>
      <td>-0.749998</td>
    </tr>
    <tr>
      <th>metagene 4</th>
      <td>1.270096</td>
      <td>-0.606707</td>
      <td>-0.327726</td>
      <td>-1.167298</td>
      <td>0.696134</td>
      <td>4.736901</td>
      <td>0.431132</td>
      <td>-0.482832</td>
      <td>-0.696304</td>
      <td>-0.131118</td>
      <td>...</td>
      <td>0.957502</td>
      <td>0.138608</td>
      <td>0.344890</td>
      <td>2.195820</td>
      <td>0.758968</td>
      <td>1.167662</td>
      <td>1.539592</td>
      <td>1.285748</td>
      <td>0.216411</td>
      <td>-0.859679</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 21766 columns</p>
</div></div>
</div>
<p>We can check that we do obtain sparse ICA components by plotting the distribution of “metagene 1” values (i.e the coefficients associated with “metagene 1” for each feature)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">distribution</span><span class="p">(</span><span class="n">Components</span> <span class="p">,</span>  <span class="n">name</span> <span class="p">,</span> <span class="n">bins</span><span class="p">):</span>
    <span class="n">s</span> <span class="o">=</span> <span class="n">Components</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="n">s</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">bins</span> <span class="o">=</span> <span class="n">bins</span> <span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="o">+</span> <span class="mf">1.</span> <span class="o">/</span> <span class="n">s</span><span class="o">.</span><span class="n">size</span> <span class="p">,</span> <span class="n">grid</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">10</span> <span class="p">,</span> <span class="mi">7</span><span class="p">))</span>
<span class="n">distribution</span><span class="p">(</span><span class="n">Metagenes</span> <span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;metagene 1&#39;</span> <span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">30</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Distribution for metagene 1&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Coefficient value&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Frequency&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Text(0, 0.5, &#39;Frequency&#39;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/_examples_transcriptomic_ICA_17_1.png" src="../_images/_examples_transcriptomic_ICA_17_1.png" />
</div>
</div>
</section>
<section id="3.-Metagene-annotation">
<h2>3. Metagene annotation<a class="headerlink" href="#3.-Metagene-annotation" title="Permalink to this headline">¶</a></h2>
<p>Finally, we provide tools to perform enrichment analysis and characterize each metagene with relevant biological annotations. These tools gather two major operations : first, a list of the most expressed genes is retrieved for each metagene, and then an enrichment analysis is performed on this list using <a class="reference external" href="https://reactome.org/">reactome</a> or <a class="reference external" href="https://toppgene.cchmc.org/enrichment.jsp">ToppFun</a>.</p>
<p>A conversion tool (<a class="reference external" href="https://pypi.org/project/mygene/">https://pypi.org/project/mygene/</a>) is included in both methods. If the user specifies the input type (e.g ‘symbol’ , ‘ensembl.gene’ , ‘uniprot’…), gene IDs will automatically be converted to Entrez gene IDs. For ToppFun this conversion is necessary for the analysis to work.</p>
<section id="3.1-Reactome-enrichment-analysis">
<h3>3.1 Reactome enrichment analysis<a class="headerlink" href="#3.1-Reactome-enrichment-analysis" title="Permalink to this headline">¶</a></h3>
<p><strong>Note :</strong> This tool is based on the python package <a class="reference external" href="https://pypi.org/project/reactome2py/">reactome2py</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">sica.annotate</span> <span class="kn">import</span> <span class="n">reactome</span>
</pre></div>
</div>
</div>
<p>The user can choose different strategies for selecting the most expressed genes of each metagene : - The user can choose to only look at the right tail of the metagene distribution, the left tail, both or the heaviest one. - The user can choose whether the thresholds will be computed with deviations from the mean of the distribution (i.e <span class="math notranslate nohighlight">\(\mu \pm k*\sigma\)</span>) or with its quantiles. - The user can fine tune these thresholds with the threshold parameter.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">Rannot</span> <span class="o">=</span> <span class="n">reactome</span><span class="o">.</span><span class="n">ReactomeAnalysis</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">Metagenes</span> <span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="mi">3</span> <span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;std&#39;</span> <span class="p">,</span> <span class="n">tail</span> <span class="o">=</span> <span class="s1">&#39;heaviest&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\ncaptier\Documents\Python\Stabilized_ICA_local\sica\annotate\reactome.py:57: UserWarning: If input_type is None the conversion of input IDs to Entrez IDs will not be possible.The inputs should be valid IDs accepted by the reactome analysis tool ((e.g HUGO gene symbols, EntrezGene , Uniprot ...)
  warnings.warn(&#34;If input_type is None the conversion of input IDs to Entrez IDs will not be possible.The inputs should be valid IDs accepted by the reactome analysis tool ((e.g HUGO gene symbols, EntrezGene , Uniprot ...)&#34;)
</pre></div></div>
</div>
<p><strong>Note :</strong> By default, an arbitrary threshold of 0.05 is used to filter Entities pValue. The user can change it with the p_value parameter. The user can also pass a list of species to filter the results as well as min and max thresholds for the number of entities per pathway.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">Rannot</span><span class="o">.</span><span class="n">get_analysis</span><span class="p">(</span><span class="n">metagene</span> <span class="o">=</span> <span class="s1">&#39;metagene 7&#39;</span> <span class="p">,</span> <span class="n">species</span> <span class="o">=</span> <span class="s1">&#39;Homo sapiens&#39;</span> <span class="p">,</span> <span class="n">p_value</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Pathway identifier</th>
      <th>Pathway name</th>
      <th>#Entities found</th>
      <th>#Entities total</th>
      <th>Entities ratio</th>
      <th>Entities pValue</th>
      <th>Entities FDR</th>
      <th>#Reactions found</th>
      <th>#Reactions total</th>
      <th>Reactions ratio</th>
      <th>Species identifier</th>
      <th>Species name</th>
      <th>Submitted entities found</th>
      <th>Mapped entities</th>
      <th>Found reaction identifiers</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>R-HSA-380108</td>
      <td>Chemokine receptors bind chemokines</td>
      <td>16</td>
      <td>57</td>
      <td>0.00392508</td>
      <td>1.8705e-11</td>
      <td>1.15784e-08</td>
      <td>12</td>
      <td>19</td>
      <td>0.00141516</td>
      <td>9606</td>
      <td>Homo sapiens</td>
      <td>6373;4283;6375;6352;6363;729230;6366;6356;2833...</td>
      <td></td>
      <td>R-HSA-373061;R-HSA-373073;R-HSA-443986;R-HSA-3...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>R-HSA-198933</td>
      <td>Immunoregulatory interactions between a Lympho...</td>
      <td>30</td>
      <td>316</td>
      <td>0.0217601</td>
      <td>1.56343e-08</td>
      <td>3.22066e-06</td>
      <td>24</td>
      <td>44</td>
      <td>0.00327722</td>
      <td>9606</td>
      <td>Homo sapiens</td>
      <td>89790;22914;3683;3385;56253;23547;10219;57823;...</td>
      <td></td>
      <td>R-HSA-199043;R-HSA-198983;R-HSA-199079;R-HSA-1...</td>
    </tr>
    <tr>
      <th>5</th>
      <td>R-HSA-202433</td>
      <td>Generation of second messenger molecules</td>
      <td>12</td>
      <td>59</td>
      <td>0.0040628</td>
      <td>1.97559e-07</td>
      <td>2.42997e-05</td>
      <td>9</td>
      <td>17</td>
      <td>0.0012662</td>
      <td>9606</td>
      <td>Homo sapiens</td>
      <td>919;3117;9402;3123;5336;7535;915;3932;3702;916...</td>
      <td></td>
      <td>R-HSA-202241;R-HSA-202375;R-HSA-202407;R-HSA-2...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Finally, the complete analysis can be browsed and visualized in reactome web portal (you need an internet access for this function).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">Rannot</span><span class="o">.</span><span class="n">open_full_analysis</span><span class="p">(</span><span class="n">metagene</span> <span class="o">=</span> <span class="s1">&#39;metagene 7&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="3.2-Toppfun-enrichment-analysis">
<h3>3.2 Toppfun enrichment analysis<a class="headerlink" href="#3.2-Toppfun-enrichment-analysis" title="Permalink to this headline">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">sica.annotate</span> <span class="kn">import</span> <span class="n">toppfun</span>
</pre></div>
</div>
</div>
<p><strong>Note :</strong> ToppFunAnalysis intializes the same way as ReactomeAnalysis. Please refer to the precedent section for some explanations.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">Tannot</span> <span class="o">=</span> <span class="n">toppfun</span><span class="o">.</span><span class="n">ToppFunAnalysis</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">Metagenes</span> <span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="mi">3</span> <span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;std&#39;</span> <span class="p">,</span> <span class="n">tail</span> <span class="o">=</span> <span class="s1">&#39;heaviest&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\ncaptier\Documents\Python\Stabilized_ICA_local\sica\annotate\toppfun.py:56: UserWarning: If input_type is None the conversion of input IDs to Entrez ids will not be possible. ToppFunAnalysis will assume that the inputs are already Entrez IDs.
  warnings.warn(&#34;If input_type is None the conversion of input IDs to Entrez ids will not be possible. ToppFunAnalysis will assume that the inputs are already Entrez IDs.&#34;)
</pre></div></div>
</div>
<p><strong>Note :</strong> By default, an arbitrary threshold of 0.05 is used to filter pvalues. The user can change it with the p_value parameter. The user can also choose which annotation features he wants to look at with the type_list parameter. All the available features are enriched by default. Finaly, the user can play with several other parameters to filter the results (e.g min_entities , max_entities…).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">Tannot</span><span class="o">.</span><span class="n">get_analysis</span><span class="p">(</span><span class="n">metagene</span> <span class="o">=</span> <span class="s1">&#39;metagene 7&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Enrichment analysis success!
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Category</th>
      <th>ID</th>
      <th>Name</th>
      <th>PValue</th>
      <th>QValueFDRBH</th>
      <th>QValueFDRBY</th>
      <th>QValueBonferroni</th>
      <th>TotalGenes</th>
      <th>GenesInTerm</th>
      <th>GenesInQuery</th>
      <th>GenesInTermInQuery</th>
      <th>Source</th>
      <th>URL</th>
      <th>Gene_Symbol</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>GeneOntologyMolecularFunction</td>
      <td>GO:0140375</td>
      <td>immune receptor activity</td>
      <td>3.436243e-20</td>
      <td>1.597853e-17</td>
      <td>1.073810e-16</td>
      <td>1.597853e-17</td>
      <td>19415</td>
      <td>141</td>
      <td>351</td>
      <td>27</td>
      <td></td>
      <td></td>
      <td>IL10RA,IL12B,IL12RB1,HLA-DOB,HLA-DQA1,HLA-DRB1...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GeneOntologyMolecularFunction</td>
      <td>GO:0004896</td>
      <td>cytokine receptor activity</td>
      <td>9.187385e-15</td>
      <td>2.136067e-12</td>
      <td>1.435507e-11</td>
      <td>4.272134e-12</td>
      <td>19415</td>
      <td>97</td>
      <td>351</td>
      <td>19</td>
      <td></td>
      <td></td>
      <td>IL10RA,IL12B,IL12RB1,IL18RAP,GFRA1,CXCR5,CCR2,...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>GeneOntologyMolecularFunction</td>
      <td>GO:0019955</td>
      <td>cytokine binding</td>
      <td>3.755193e-14</td>
      <td>5.820549e-12</td>
      <td>3.911600e-11</td>
      <td>1.746165e-11</td>
      <td>19415</td>
      <td>184</td>
      <td>351</td>
      <td>24</td>
      <td></td>
      <td></td>
      <td>IL10RA,IL12B,IL12RB1,TNFRSF9,NLRP7,GBP1,IL18RA...</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>For both methods (i.e ReactomeAnalysis and ToppFunAnalysis) the top_genes_ attribute gives access to the list of extreme expressed genes for each metagene. The ‘inputs’ column contains the original gene IDs, the ‘entrezgene’ column contains the converted gene IDs (in Entrez IDs), and the ‘notfound’ column contains the original gene IDs which were not found by the conversion tool.</p>
<p><strong>Note :</strong> For cases with no conversion( i.e input_type parameter was set to None), ‘entregene’ and ‘notfound’ columns contain only None values.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">Tannot</span><span class="o">.</span><span class="n">top_genes_</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>inputs</th>
      <th>entrezgene</th>
      <th>notfound</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>metagene 0</th>
      <td>[6192, 9086, 8653, 8284, 8287, 246126, 7404, 6...</td>
      <td>None</td>
      <td>None</td>
    </tr>
    <tr>
      <th>metagene 1</th>
      <td>[128602, 92747, 399949, 127003, 144448, 133690...</td>
      <td>None</td>
      <td>None</td>
    </tr>
    <tr>
      <th>metagene 2</th>
      <td>[388468, 400968, 29044, 228, 160857, 83417, 72...</td>
      <td>None</td>
      <td>None</td>
    </tr>
    <tr>
      <th>metagene 3</th>
      <td>[79832, 1038, 23347, 7403, 83448, 659, 163590,...</td>
      <td>None</td>
      <td>None</td>
    </tr>
    <tr>
      <th>metagene 4</th>
      <td>[283131, 79923, 29896, 84719, 27099, 100642175...</td>
      <td>None</td>
      <td>None</td>
    </tr>
    <tr>
      <th>metagene 5</th>
      <td>[1301, 131578, 6423, 8483, 26585, 84624, 22801...</td>
      <td>None</td>
      <td>None</td>
    </tr>
    <tr>
      <th>metagene 6</th>
      <td>[8685, 1118, 366, 81501, 126014, 6348, 344, 11...</td>
      <td>None</td>
      <td>None</td>
    </tr>
    <tr>
      <th>metagene 7</th>
      <td>[260436, 931, 10563, 83416, 3507, 973, 608, 13...</td>
      <td>None</td>
      <td>None</td>
    </tr>
    <tr>
      <th>metagene 8</th>
      <td>[361, 653509, 6440, 1510, 729238, 51208, 1755,...</td>
      <td>None</td>
      <td>None</td>
    </tr>
    <tr>
      <th>metagene 9</th>
      <td>[9787, 83540, 55872, 4751, 259266, 55635, 1063...</td>
      <td>None</td>
      <td>None</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
</section>
</section>


      </div>
    <div class="container">
      <footer class="sk-content-footer">
            &copy; 2007-2021 The scikit-learn developers (BSD License).
          <a href="../_sources/_examples/transcriptomic_ICA.ipynb.txt" rel="nofollow">Show this page source</a>
      </footer>
    </div>
  </div>
</div>
<script src="../_static/js/vendor/bootstrap.min.js"></script>

<script>
    window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
    ga('create', 'UA-22606712-2', 'auto');
    ga('set', 'anonymizeIp', true);
    ga('send', 'pageview');
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>


<script>
$(document).ready(function() {
    /* Add a [>>>] button on the top-right corner of code samples to hide
     * the >>> and ... prompts and the output and thus make the code
     * copyable. */
    var div = $('.highlight-python .highlight,' +
                '.highlight-python3 .highlight,' +
                '.highlight-pycon .highlight,' +
		'.highlight-default .highlight')
    var pre = div.find('pre');

    // get the styles from the current theme
    pre.parent().parent().css('position', 'relative');
    var hide_text = 'Hide prompts and outputs';
    var show_text = 'Show prompts and outputs';

    // create and add the button to all the code blocks that contain >>>
    div.each(function(index) {
        var jthis = $(this);
        if (jthis.find('.gp').length > 0) {
            var button = $('<span class="copybutton">&gt;&gt;&gt;</span>');
            button.attr('title', hide_text);
            button.data('hidden', 'false');
            jthis.prepend(button);
        }
        // tracebacks (.gt) contain bare text elements that need to be
        // wrapped in a span to work with .nextUntil() (see later)
        jthis.find('pre:has(.gt)').contents().filter(function() {
            return ((this.nodeType == 3) && (this.data.trim().length > 0));
        }).wrap('<span>');
    });

    // define the behavior of the button when it's clicked
    $('.copybutton').click(function(e){
        e.preventDefault();
        var button = $(this);
        if (button.data('hidden') === 'false') {
            // hide the code output
            button.parent().find('.go, .gp, .gt').hide();
            button.next('pre').find('.gt').nextUntil('.gp, .go').css('visibility', 'hidden');
            button.css('text-decoration', 'line-through');
            button.attr('title', show_text);
            button.data('hidden', 'true');
        } else {
            // show the code output
            button.parent().find('.go, .gp, .gt').show();
            button.next('pre').find('.gt').nextUntil('.gp, .go').css('visibility', 'visible');
            button.css('text-decoration', 'none');
            button.attr('title', hide_text);
            button.data('hidden', 'false');
        }
    });

	/*** Add permalink buttons next to glossary terms ***/
	$('dl.glossary > dt[id]').append(function() {
		return ('<a class="headerlink" href="#' +
			    this.getAttribute('id') +
			    '" title="Permalink to this term">¶</a>');
	});
  /*** Hide navbar when scrolling down ***/
  // Returns true when headerlink target matches hash in url
  (function() {
    hashTargetOnTop = function() {
        var hash = window.location.hash;
        if ( hash.length < 2 ) { return false; }

        var target = document.getElementById( hash.slice(1) );
        if ( target === null ) { return false; }

        var top = target.getBoundingClientRect().top;
        return (top < 2) && (top > -2);
    };

    // Hide navbar on load if hash target is on top
    var navBar = document.getElementById("navbar");
    var navBarToggler = document.getElementById("sk-navbar-toggler");
    var navBarHeightHidden = "-" + navBar.getBoundingClientRect().height + "px";
    var $window = $(window);

    hideNavBar = function() {
        navBar.style.top = navBarHeightHidden;
    };

    showNavBar = function() {
        navBar.style.top = "0";
    }

    if (hashTargetOnTop()) {
        hideNavBar()
    }

    var prevScrollpos = window.pageYOffset;
    hideOnScroll = function(lastScrollTop) {
        if (($window.width() < 768) && (navBarToggler.getAttribute("aria-expanded") === 'true')) {
            return;
        }
        if (lastScrollTop > 2 && (prevScrollpos <= lastScrollTop) || hashTargetOnTop()){
            hideNavBar()
        } else {
            showNavBar()
        }
        prevScrollpos = lastScrollTop;
    };

    /*** high performance scroll event listener***/
    var raf = window.requestAnimationFrame ||
        window.webkitRequestAnimationFrame ||
        window.mozRequestAnimationFrame ||
        window.msRequestAnimationFrame ||
        window.oRequestAnimationFrame;
    var lastScrollTop = $window.scrollTop();

    if (raf) {
        loop();
    }

    function loop() {
        var scrollTop = $window.scrollTop();
        if (lastScrollTop === scrollTop) {
            raf(loop);
            return;
        } else {
            lastScrollTop = scrollTop;
            hideOnScroll(lastScrollTop);
            raf(loop);
        }
    }
  })();
});

</script>
    
</body>
</html>